staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+
+	gwebsocket "github.com/gorilla/websocket"
+
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+	"k8s.io/apimachinery/pkg/util/httpstream/wsstream"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	constants "k8s.io/apimachinery/pkg/util/portforward"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/klog/v2"
+)
+
+// TunnelingHandler is a handler which tunnels SPDY through WebSockets.
+type TunnelingHandler struct {
+	// UpgradeAwareHandler used to communicate to upstream SPDY.
+	upgradeHandler http.Handler
+	// Server-side of upgraded websocket connection
+	tunnelingConnection net.Conn
+	// MaxBytesPerSec throttles stream Reader/Writer if necessary
+	MaxBytesPerSec int64
+}
+
+func NewTunnelingHandler(upgradeHandler http.Handler, maxBytesPerSec int64) *TunnelingHandler {
+	return &TunnelingHandler{
+		upgradeHandler: upgradeHandler,
+		MaxBytesPerSec: maxBytesPerSec,
+	}
+}
+
+func (h *TunnelingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	klog.Infoln("TunnelingHandler ServeHTTP")
+	clone := utilnet.CloneRequest(req)
```

Comment:
might need to nil out or set the body of the cloned request to a Noop reader... we don't expect the body of the request to be read at all by the kubelet, only the hijacked connection

Reply:
Fixed.